### PR TITLE
[MNOE-426] - Logged additional events

### PR DESCRIPTION
**Pull request by fgourichon**
@ouranos for review

### CODE REVIEW

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 1]
```diff
@@ -25,7 +25,12 @@ def index
   def create
     if widgets
       if @widget = widgets.create(widget_create_params)
-        MnoEnterprise::EventLogger.info('widget_create', current_user.id, 'Widget Creation', @widget)
+        if @widget.settings['organization_ids'].length >= 1
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: There few operations like this one along the PR. 
Would not be more clear something like?
if you are expecting an array
```
if @widget.settings['organization_ids'].any?
```
or if you are expecting a string
```
if @widget.settings['organization_ids'].present?
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 2]
```diff
@@ -40,6 +45,12 @@ def create
   #   -> PUT /api/mnoe/v1/widgets/:id
   def update
     if widget.update(widget_update_params)
+      if @widget.settings['organization_ids'].length >= 1
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: same as before?
```
if @widget.settings['organization_ids'].any?
```
```
if @widget.settings['organization_ids'].present?
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 3]
```diff
@@ -51,7 +62,13 @@ def update
   #   -> DELETE /api/mnoe/v1/widgets/:id
   def destroy
     if widget.destroy
-      MnoEnterprise::EventLogger.info('widget_delete', current_user.id, 'Widget Deletion', widget)
+      if @widget.settings['organization_ids'].length >= 1
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: same as before?
```
if @widget.settings['organization_ids'].any?
```
```
if @widget.settings['organization_ids'].present?
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 4]
```diff
@@ -25,7 +25,12 @@ def index
   def create
     if widgets
       if @widget = widgets.create(widget_create_params)
-        MnoEnterprise::EventLogger.info('widget_create', current_user.id, 'Widget Creation', @widget)
+        if @widget.settings['organization_ids'].length >= 1
+          organization = MnoEnterprise::Organization.find_by(uid: @widget.settings['organization_ids'][0])
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: What about?
```
@widget.settings['organization_ids'].first
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 5]
```diff
@@ -40,6 +45,12 @@ def create
   #   -> PUT /api/mnoe/v1/widgets/:id
   def update
     if widget.update(widget_update_params)
+      if @widget.settings['organization_ids'].length >= 1
+        organization = MnoEnterprise::Organization.find_by(uid: @widget.settings['organization_ids'][0])
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: same as before?
```
@widget.settings['organization_ids'].first
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 6]
```diff
@@ -51,7 +62,13 @@ def update
   #   -> DELETE /api/mnoe/v1/widgets/:id
   def destroy
     if widget.destroy
-      MnoEnterprise::EventLogger.info('widget_delete', current_user.id, 'Widget Deletion', widget)
+      if @widget.settings['organization_ids'].length >= 1
+        organization = MnoEnterprise::Organization.find_by(uid: @widget.settings['organization_ids'][0])
```
Comment (hedudelgado) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: same as before?
```
@widget.settings['organization_ids'].first
```

**Comment by fgourichon** [comment]
Thx @hedudelgado !

**Comment by fgourichon** [comment]
@ouranos you can have a look, I moved logic in to_audit_event for widgets, for the others extra parameters are contextual to the call so I left them as additional metadata.

I also merged existing metadata with provided ones rather than replacing.

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb [hunk 7]
```diff
@@ -30,6 +30,8 @@ def create
     authorize! :manage_teams, parent_organization
     @team = parent_organization.teams.create(team_params)
 
+    MnoEnterprise::EventLogger.info('team_add', current_user.id, 'Team created', parent_organization, {team_name: @team.name}) if @team
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]: `{team_name: @team,name}` could be moved to the model `to_audit_event` as well as the `organization_id`?
So we'd have
```ruby
MnoEnterprise::EventLogger.info('team_add', current_user.id, 'Team created', @team) if @team
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb [hunk 8]
```diff
@@ -45,6 +47,10 @@ def update
     if params[:team] && params[:team][:app_instances]
       list = params[:team][:app_instances].select { |e| e != {} }
       @team.set_access_to(list)
+
+      MnoEnterprise::EventLogger.info('team_apps_update', current_user.id, 'Team apps updated', @team.organization,
+                                      {team_name: @team.name, apps: list.map{|l| l['name']}.join(',')})
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]: For the same output you could use `Array#to_sentence`
But in this case I don't mind just having:
```ruby
MnoEnterprise::EventLogger.info('team_apps_update', current_user.id, 'Team apps updated', @team, {apps: list.map {|a| a['name']}
```
The `metadata` is hash and that leave us more flexibility to format the array however we want

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb [hunk 9]
```diff
@@ -64,8 +70,11 @@ def remove_users
   def destroy
     @team = MnoEnterprise::Team.find(params[:id])
     authorize! :manage_teams, @team.organization
+
     @team.destroy
 
+    MnoEnterprise::EventLogger.info('team_delete', current_user.id, 'Team deleted', @team.organization, {team_name: @team.name}) if @team
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]:
```ruby
MnoEnterprise::EventLogger.info('team_delete', current_user.id, 'Team deleted', @team) if @maestrano/engineering 
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb [hunk 10]
```diff
@@ -80,7 +89,14 @@ def update_members(action)
...
       id_list = params[:team][:users].map { |h| h[:id] }.compact
       users = @team.organization.users.where('id.in' => id_list)
-      users.each { |u| @team.send(action, u) }
+
+      user_list = users.map do |u|
+        @team.send(action, u)
+        u.email
+      end.join(',')
+
+      MnoEnterprise::EventLogger.info('team_update', current_user.id, 'Team composition updated', @team.organization,
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]:
```ruby
users.each { |u| @team.send(action, u) }

MnoEnterprise::EventLogger.info('team_update', current_user.id, 'Team composition updated', @team, {action: action.to_s,  users: users.map(&:email)
```

Or you could keep the `map` as you had, it's sligthly more efficent but I think it's harder to read? No strong opinion on this one

File: api/lib/mno_enterprise/concerns/controllers/webhook/o_auth_controller.rb [hunk 11]
```diff
@@ -76,6 +76,15 @@ def callback
       path = add_param_to_fragment(path.to_s, 'flash', [{msg: error_message(error_key),  type: :error}.to_json])
     end
 
+    unless params.fetch(:oauth, {})[:error]
+      case params.fetch(:oauth, {})[:action]
+        when 'sync'
+          MnoEnterprise::EventLogger.info('app_connected', current_user.id, 'App connected', app_instance.owner, {app: app_instance.name})
+        when 'disconnect'
+          MnoEnterprise::EventLogger.info('app_disconnected', current_user.id, 'App diconnected', app_instance.owner, {app: app_instance.name})
+      end
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/webhook/o_auth_controller.rb]:
```ruby
MnoEnterprise::EventLogger.info('app_disconnected', current_user.id, 'App diconnected', app_instance)
```
https://github.com/maestrano/mno-enterprise/blob/6039f43129e6daedfc60aad9652dca0fc7939896/core/lib/mno_enterprise/concerns/models/app_instance.rb#L104-L112

File: core/lib/mno_enterprise/concerns/controllers/auth/registrations_controller.rb [hunk 12]
```diff
@@ -46,10 +46,13 @@ module ClassMethods
...
     build_resource(sign_up_params)
     resource.password ||= Devise.friendly_token
-    
+
     resource_saved = resource.save
-    
+
     if resource_saved
+
+      MnoEnterprise::EventLogger.info('user_add', resource_saved.id, 'User Signup', OpenStruct.new(id: 0 ), {email: resource_saved.email})
```
Comment (ouranos) [inline core/lib/mno_enterprise/concerns/controllers/auth/registrations_controller.rb]: Wouldn't `resource` be an instance of `MnoEnterprise::User`?
If so you could use:
```ruby
MnoEnterprise::EventLogger.info('user_add', resource_saved.id, 'User Signup', resource_saved)
```

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 13]
```diff
@@ -51,7 +52,7 @@ def update
   #   -> DELETE /api/mnoe/v1/widgets/:id
   def destroy
     if widget.destroy
-      MnoEnterprise::EventLogger.info('widget_delete', current_user.id, 'Widget Deletion', widget)
+      MnoEnterprise::EventLogger.info('widget_delete', current_user.id, 'Widget Deletion', @widget)
```
Comment (fgourichon) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: I guess it needs to be changed everywhere in the controller?

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb (hunk 7 above)
Comment (fgourichon) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]: Yes

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb (hunk 8 above)
Comment (fgourichon) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]: Done

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb (hunk 9 above)
Comment (fgourichon) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/teams_controller.rb]: Team seeked and destroyed.

**Comment by fgourichon** [comment]
@ouranos ready for final review

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 14]
```diff
@@ -17,19 +17,19 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::WidgetsController
   #  -> GET /api/mnoe/v1/organizations/:id/widgets
   def index
     render_not_found('organization') unless parent_organization
-    @widgets = parent_organization.widgets
+    widgets = parent_organization.widgets
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: This need to be an ivar to be shared with the view (see the specs failing)

File: api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb [hunk 15]
```diff
@@ -17,19 +17,19 @@ module MnoEnterprise::Concerns::Controllers::Jpi::V1::Impac::WidgetsController
...
   end
 
   # POST /mnoe/jpi/v1/impac/dashboards/:id/widgets
   #  -> POST /api/mnoe/v1/dashboards/:id/widgets
   def create
     if widgets
-      if @widget = widgets.create(widget_create_params)
-        MnoEnterprise::EventLogger.info('widget_create', current_user.id, 'Widget Creation', @widget)
+      if widget = widgets.create(widget_create_params)
```
Comment (ouranos) [inline api/lib/mno_enterprise/concerns/controllers/jpi/v1/impac/widgets_controller.rb]: Needs to be an ivar

**Comment by fgourichon** [comment]
@ouranos ivar reverted

**Pull request merged**